include/mesh.tcc

```diff
@@ -299,22 +298,104 @@ void mpm::Mesh<Tdim>::compute_cell_neighbours() {
   }
 
   // Assign neighbour to cells
-  tbb::parallel_for(
-      tbb::blocked_range<int>(size_t(0), size_t(cells_.size()),
-                              tbb_grain_size_),
-      [&](const tbb::blocked_range<int>& range) {
-        for (int i = range.begin(); i != range.end(); ++i) {
-          // Iterate over each node in current cell
-          for (auto id : cells_[i]->nodes_id()) {
-            auto cell_id = cells_[i]->id();
-            // Get the cells associated with each node
-            for (auto neighbour_id : node_cell_map[id])
-              if (neighbour_id != cell_id)
-                cells_[i]->add_neighbour(neighbour_id);
-          }
-        }
-      },
-      tbb::simple_partitioner());
+  tbb::parallel_for(tbb::blocked_range<int>(size_t(0), size_t(cells_.size()),
+                                            tbb_grain_size_),
+                    [&](const tbb::blocked_range<int>& range) {
+                      for (int i = range.begin(); i != range.end(); ++i) {
+                        // Iterate over each node in current cell
+                        for (auto id : cells_[i]->nodes_id()) {
+                          auto cell_id = cells_[i]->id();
+                          // Get the cells associated with each node
+                          for (auto neighbour_id : node_cell_map[id])
+                            if (neighbour_id != cell_id)
+                              cells_[i]->add_neighbour(neighbour_id);
+                        }
+                      }
+                    },
+                    tbb::simple_partitioner());
+}
+
+//! Find particle neighbours for all particle
+template <unsigned Tdim>
+void mpm::Mesh<Tdim>::compute_particle_neighbours() {
+  // Check whether cells have been assigned with neighbours
+  if (!(*cells_.cbegin())->has_neighbours()) this->compute_cell_neighbours();
+
+  // Loop over cells
+  for (auto citr = cells_.cbegin(); citr != cells_.cend(); ++citr) {
+    // Loop over the neighbouring cell particles
+    std::vector<mpm::Index> neighbouring_particle_sets = (*citr)->particles();
```

Comment:
avoid using type names. This is a vector and not a set. Please call it `neighbour_particles`